### PR TITLE
ARO-24037: gate cloud config on hash to prevent serving stale content

### PR DESCRIPTION
**Pull request by twolff-gh**
Prevents stale cloud.conf from being served during the race between the CPO updating azure-cloud-config and the ignition server reading it.

Follows the existing pullSecretHash / additionalTrustBundleHash pattern: nodepool controller hashes the cloud config ConfigMap and stores it in the token secret; ignition server verifies the hash before writing cloud.conf, retrying on mismatch.

## Which issue(s) this PR fixes:

Fixes https://redhat.atlassian.net/browse/ARO-24037

Related to https://github.com/openshift/hypershift/pull/8865

## Checklist:
- [ ] Subject and description added to both, commit and PR.
- [ ] Relevant issues have been referenced.
- [ ] This change includes docs. 
- [ ] This change includes unit tests.

## Summary by CodeRabbit

* **New Features**
  * Added tracking of cloud provider configuration via a new `cloud-config-hash` value stored in token Secrets and propagated into ignition payload generation.
  * Ignition payload caching/reuse now records and considers this cloud configuration hash.

* **Bug Fixes**
  * Azure/OpenStack ignition payloads now validate the retrieved cloud configuration content against the expected hash, preventing stale or out-of-date configuration.
  * Token secret reconciliation and payload cache invalidation are improved: payloads are regenerated when the cloud configuration hash changes.

### CODE REVIEW

File: hypershift-operator/controllers/nodepool/token.go [hunk 1]
```diff
@@ -354,11 +362,13 @@ func (t *Token) reconcileTokenSecret(tokenSecret *corev1.Secret) error {
 		tokenSecret.Data[TokenSecretPullSecretHashKey] = t.pullSecretHash
 		tokenSecret.Data[TokenSecretAdditionalTrustBundleKey] = t.additionalTrustBundleHash
 		tokenSecret.Data[TokenSecretHCConfigurationHashKey] = t.globalConfigHash
+		tokenSecret.Data[TokenSecretCloudConfigHashKey] = t.cloudConfigHash
 	}
 	// TODO (alberto): Only apply this on creation and change the hash generation to only use triggering upgrade fields.
 	// We let this change to happen inplace now as the tokenSecret and the mcs config use the whole spec.Config for the comparing hash.
 	// Otherwise if something which does not trigger a new token generation from spec.Config changes, like .IDP, both hashes would mismatch forever.
 	tokenSecret.Data[TokenSecretHCConfigurationHashKey] = t.globalConfigHash
+	tokenSecret.Data[TokenSecretCloudConfigHashKey] = t.cloudConfigHash
```
[the remark this answers is not in the document]
Comment (enxebre) [reply hypershift-operator/controllers/nodepool/token.go:375]: agree, let's drop this line

**Comment by jparrill** [comment]
/retitle: [ARO-24037](https://redhat.atlassian.net/browse/ARO-24037): gate cloud config on hash to prevent serving stale content

**Review by jparrill — COMMENTED** [review]
Dropped some comments. Thanks!

I've retitled the issue in order to fill the bot expectations

File: hypershift-operator/controllers/nodepool/token.go [hunk 2]
```diff
@@ -113,13 +115,19 @@ func NewToken(ctx context.Context, configGenerator *ConfigGenerator, cpoCapabili
...
+	cloudConfigHash, err := configGenerator.GetCloudConfigHash(ctx)
+	if err != nil {
+		return nil, err
+	}
+
 	token := &Token{
 		CreateOrUpdateProvider:    upsert.New(false),
 		ConfigGenerator:           configGenerator,
 		cpoCapabilities:           cpoCapabilities,
 		pullSecretHash:            []byte(supportutil.HashSimple(pullSecretBytes)),
```
Comment (jparrill) [inline hypershift-operator/controllers/nodepool/token.go:128]: There's a convergence problem here. `ConfigGenerator.Hash()` doesn't include `cloudConfigHash`, so `token.isOutdated()` won't return true when cloud config changes — no new token gets created. This in-place update is the only mechanism that propagates the new hash to existing token secrets.

But the nodepool controller has no watch on cloud config ConfigMaps. `enqueueNodePoolsForConfig` only matches ConfigMaps with the `core-ignition-config` label (cloud config CMs don't have it), and the handler lists NodePools in `cm.Namespace` — cloud config CMs are in the CP namespace, NodePools in the management namespace. After CPO updates the ConfigMap, the nodepool controller never re-reconciles, this line never executes, and the ignition server retries indefinitely.

The fix needs a convergence mechanism — whether that's adding the hash to `ConfigGenerator.Hash()`, adding a watch, or labeling the ConfigMaps is your call.

File: support/util/util.go [hunk 3]
```diff
@@ -178,6 +178,24 @@ func InsecureHTTPClient() *http.Client {
 	}
```
Comment (jparrill) [inline support/util/util.go:178]: Missing dedicated unit tests — nil map, empty map, determinism (same keys different insertion order), and delimiter correctness (`{"ab": "c"}` vs `{"a": "bc"}`).

Also: if `cm.Data` is nil (valid k8s state), this returns a hash of empty string, which differs from the `""` returned by `GetCloudConfigHash` when the ConfigMap is NotFound. Intentional?

File: ignition-server/controllers/local_ignitionprovider.go [hunk 4]
```diff
@@ -300,8 +300,10 @@ func (p *LocalIgnitionProvider) extractImageReferences(ctx context.Context, rele
...
 }
 
-// For Azure and OpenStack, extract the cloud provider config file as MCO input
-func (p *LocalIgnitionProvider) writeCloudProviderConfig(ctx context.Context, mcoDir string) error {
+// For Azure and OpenStack, extract the cloud provider config file as MCO input.
+// When cloudConfigHash is non-empty, the ConfigMap content is verified against the hash
+// before writing to ensure the ignition server does not serve stale cloud config.
+func (p *LocalIgnitionProvider) writeCloudProviderConfig(ctx context.Context, mcoDir, cloudConfigHash string) error {
 	if p.CloudProvider != hyperv1.AzurePlatform && p.CloudProvider != hyperv1.OpenStackPlatform {
 		return nil
```
Comment (jparrill) [inline ignition-server/controllers/local_ignitionprovider.go:308]: The error message should include expected and actual hash values — without them an operator can't distinguish transient race from permanent mismatch:
```go
return fmt.Errorf("cloud config %s/%s hash mismatch (expected %s, got %s), waiting for CPO update",
    cloudConfigMap.Namespace, cloudConfigMap.Name, cloudConfigHash, util.HashConfigMapData(cloudConfigMap.Data))
```

File: ignition-server/controllers/tokensecret_controller.go [hunk 5]
```diff
@@ -226,23 +227,29 @@ func (r *TokenSecretReconciler) Reconcile(ctx context.Context, req ctrl.Request)
 	}
 
 	token := string(tokenSecret.Data[TokenSecretTokenKey])
+	cloudConfigHash := string(tokenSecret.Data[TokenSecretCloudConfigHashKey])
 	if value, ok := r.PayloadStore.Get(token); ok {
```
Comment (jparrill) [inline ignition-server/controllers/tokensecret_controller.go:237]: One minor asymmetry: for the primary token, you explicitly `Delete(token)` from the cache on mismatch, but for the old token (line ~244), the stale entry is just skipped without being deleted.

Also, `InvalidConfigReason` for the mismatch error is semantically misleading — the config isn't invalid, it's stale/pending. Operators monitoring for `InvalidConfig` might confuse a transient race with a broken configuration. Consider a distinct reason like `"CloudConfigPending"`.

File: hypershift-operator/controllers/nodepool/config_test.go [hunk 6]
```diff
@@ -12,6 +12,7 @@ import (
```
Comment (jparrill) [inline hypershift-operator/controllers/nodepool/config_test.go:13]: Two gaps:
- Missing OpenStack test case — there's an explicit `case hyperv1.OpenStackPlatform:` branch with zero coverage.
- Missing non-NotFound error path test — the `apierrors.IsNotFound` guard returns `("", nil)` but no test verifies the wrapped error on other failures.

File: ignition-server/controllers/tokensecret_controller_test.go [hunk 7]
```diff
@@ -30,7 +30,7 @@ var (
```
Comment (jparrill) [inline ignition-server/controllers/tokensecret_controller_test.go:30]: Missing edge case: empty-to-non-empty transition (`cachedHash: "", secretHash: "new-hash"` — first-time cloud config addition).

The old-token fallback with hash mismatch (the `&& value.CloudConfigHash == cloudConfigHash` condition at line ~244) also has no test coverage.

File: support/util/util.go (hunk 3 above)
Comment (twolff-gh) [reply support/util/util.go:178]: Tests added.

Other: not intentional. Thanks for noticing it :pray: 
I think I addressed it. Lmk if it make sense

File: ignition-server/controllers/tokensecret_controller.go (hunk 5 above)
Comment (twolff-gh) [reply ignition-server/controllers/tokensecret_controller.go:237]: Addressed. Let me know if it works / makes sense

File: hypershift-operator/controllers/nodepool/config_test.go (hunk 6 above)
Comment (twolff-gh) [reply hypershift-operator/controllers/nodepool/config_test.go:13]: Addressed. Let me know if there is anything more I need to do here. :pray:

File: ignition-server/controllers/tokensecret_controller_test.go (hunk 7 above)
Comment (twolff-gh) [reply ignition-server/controllers/tokensecret_controller_test.go:30]: Added.

File: hypershift-operator/controllers/nodepool/token.go (hunk 2 above)
Comment (twolff-gh) [reply hypershift-operator/controllers/nodepool/token.go:128]: Does an in-place update + cache invalidation provide eventual convergence without rollout risk? Not instantaneous, but is that a problem here? Or should it be addressed with one of your recommended options?

On the reconciliation, wouldn't the token TTL eventually cause the cloud.conf to be updated: NewToken() calls GetCloudConfigHash() and gets the current hash from the live ConfigMap? Or even from other changed. Obviously not a great UX, and I wouldn't want that in the wild.

Adding to ConfigGenerator.Hash() would roll out to all clusters simultaneously, right? I don't think this one should be done. Unless you think it makes sense.

Does labeling the ConfigMaps require more CPO changes?

So leaning toward labeling the ConfigMaps or adding a watch. Whichever is simpler. I'll look into that.
Initially looking, adding a watch might be simpler, but would add some complexity (which I guess is unavoidable here). Totally fine to do. As part of this work though, or a separate PR?

If you have any info, lmk.

Thanks! :pray:

Comment (twolff-gh) [reply hypershift-operator/controllers/nodepool/token.go:128]: I just pushed the watch https://github.com/twolff-gh/hypershift/commit/6a66d36522dbe8c8b3e03fdadc21f826411e046f

Lmk if this is what you had in mind :pray:

**Comment by jparrill** [comment]
/retitle [ARO-24037](https://redhat.atlassian.net/browse/ARO-24037): gate cloud config on hash to prevent serving stale content

**Comment by twolff-gh** [comment]
> New changes are detected. LGTM label has been removed.
@jparrill 
Fixed merge conflicts and ran restructure-commits. Had to force push

**Comment by twolff-gh** [comment]
/verified by @twolff-gh 

Manually verified on personal dev env. Custom image arohcpsvcdev.azurecr.io/test-twolff-hypershift-operator@sha256:097be4c68028d2341850dccf86291ece85f9f27365e49a9a57d7e8de312d21a2 deployed.

Azure HostedCluster created, 2/2 nodes healthy, all ClusterOperators available. Token secret contains cloud-config-hash data key, confirming NodePool controller writes the hash and ignition server serves content successfully with hash gating enabled.
